### DISCUSSION TITLE
Error when running example " 'Equation' object has no attribute 'form'

### DISCUSSION THREAD

**Discussion by wearefor** [IvanYashchuk/jax-fenics-adjoint#17] [Q&A] [answered]
Hello! 
Excited to use this package on some control problems in fluid dynamics.
I was able to install the required dependencies but haven't been able to run any of the examples yet.

I'm attaching the error message I received when running "poisson_control.py" and a list of the packages in the conda environment.

The issue may be related to the following:
I initially installed "python -m pip install git+https://github.com/dolfin-adjoint/pyadjoint.git@master" but received the error message "No module named 'fenics_adjoint' so I also installed  "python -m pip install git+https://github.com/dolfin-adjoint/dolfin-adjoint.git@main".
Is that right?

Thanks!

[condaenv.txt](https://github.com/IvanYashchuk/jax-fenics-adjoint/files/13323793/condaenv.txt)
[errormessage.txt](https://github.com/IvanYashchuk/jax-fenics-adjoint/files/13323794/errormessage.txt)

**Comment by IvanYashchuk** [comment]
Hello Mike,

I think the `pyadjoint` project went through a bit of refactoring recently (for example this PR seems relevant https://github.com/dolfin-adjoint/pyadjoint/pull/105). I will take a look and see what changes are needed to make the examples work again with the latest pyadjoint.

**Comment by IvanYashchuk** [comment, marked as answer]
I think something got broken when the dolfin-adjoint got split from the pyadjoint repo into a separate one. With FEniCS 2019 I recommend using the following command to install pyadjoint+dolfin-adjoint: `python -m pip install git+https://github.com/dolfin-adjoint/pyadjoint.git@2023.0.0`. I'm going to update the install instructions in the README of this project.

**Reply by wearefor** [reply]
Thanks a lot, Ivan! I'm able to run the example without error now.

**Reply by IvanYashchuk** [reply]
CI is still red though because JAX has changed: https://github.com/IvanYashchuk/jax-fenics-adjoint/pull/18:
```py
tests/fenics/test_helpers.py:12: in <module>
    from jax._src.abstract_arrays import make_shaped_array
E   ImportError: cannot import name 'make_shaped_array' from 'jax._src.abstract_arrays' (/usr/share/miniconda/envs/fenicsproject/lib/python3.12/site-packages/jax/_src/abstract_arrays.py)
```
